kube-client/src/config/incluster_config.rs

```diff
@@ -29,6 +37,43 @@ pub fn kube_dns() -> http::Uri {
     http::Uri::from_static("https://kubernetes.default.svc/")
 }
 
+pub fn try_kube_from_env() -> Result<http::Uri, Error> {
+    // client-go requires that both environment variables are set.
+    let host = std::env::var("KUBERNETES_SERVICE_HOST").map_err(Error::ReadEnvironmentVariable)?;
+    let port = std::env::var("KUBERNETES_SERVICE_PORT")
+        .map_err(Error::ReadEnvironmentVariable)?
+        .parse::<u16>()
+        .map_err(Error::ParseClusterPort)?;
+
+    // Format a host and, if not using 443, a port.
+    //
+    // Ensure that IPv6 addresses are properly bracketed.
+    const HTTPS: &str = "https";
+    let uri = match host.parse::<std::net::IpAddr>() {
+        Ok(ip) => {
+            if port == 443 {
+                if ip.is_ipv6() {
+                    format!("{HTTPS}://[{ip}]")
+                } else {
+                    format!("{HTTPS}://{ip}")
+                }
+            } else {
+                let addr = std::net::SocketAddr::new(ip, port);
+                format!("{HTTPS}://{addr}")
+            }
+        }
+        Err(_) => {
+            if port == 443 {
+                format!("{HTTPS}://{host}")
+            } else {
+                format!("{HTTPS}://{host}:{port}")
+            }
+        }
+    };
```

Comment:
This is more expansive than what we had before. Appreciate you making this as nice as possible with regards to the possible other v6 issue.